src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -137,7 +137,7 @@ public bool HasListeners()
                     var getRequestedDataUsingContext = listener.GetRequestedDataUsingContext;
                     if (getRequestedDataUsingContext != null)
                     {
-                        ActivityCreationOptions<ActivityContext> aco = new ActivityCreationOptions<ActivityContext>(this, name, context, kind, tags, links);
+                        ActivityCreationOptions<ActivityContext> aco = new ActivityCreationOptions<ActivityContext>(this, name, context == default && Activity.Current != null ? Activity.Current.Context : context, kind, tags, links);
```

Comment:
Fetching Activity.Current has some perf overhead to access the async-local. Rather than repeat this operation twice per listener I'd recommend computing the parent Activity and context once outside the loop and then reusing it for all the listener callbacks and passing it into CreateAndStart.

As a sidenote not for this PR we still need to spend some more time analyzing performance. I notice the current implementation is going to allocate a closure for the delegate we are passing to EnumWithFunc and the goal is to make sampled-out Activities be a zero-allocation code-path.

Reply:
@noah 

> and passing it into CreateAndStart.

I am not going to pass it to CreateAndStart because inside we check if we set Parent property only when `Context == default`. Of course, we can check if the context equals the parent context and set the Parent property but would be good avoid that for the perf sake. anyway, we can address this in other PR, but for now, I'll apply the optimization to calculate the context once for all listeners as you suggested.

https://github.com/tarekgh/runtime/blob/95a296ede1799b1f8cdedea57451a0f51f6dabfd/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L868

Reply:
I'll address this in a separate PR as I want to merge this one to catch P6. thanks for the feedback.